FIX allows TransformedTargetRegressor to take nD target

panangam opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #18866 


#### What does this implement/fix? Explain your changes.
This removes check_array call from the fit method of TransformedTargetRegressor and relegate target validation to the transformer. The main motivation was the allow nD target to be used. But this also potentially allows transformers that deal with string or other data types to be used. 

#### Any other comments?
This is my first scikit-learn PR so any feedback welcome!

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

panangam committed:
allow_nd set to True in the check_array call

panangam committed:
simply cast y to a numpy array without validating

panangam committed:
implement a test for a 3D target

panangam committed:
shorten a line

glemaitre commented on sklearn/compose/_target.py:
```diff
@@ -175,8 +175,7 @@ def fit(self, X, y, **fit_params):
         """
-        y = check_array(y, accept_sparse=False, force_all_finite=True,
-                        ensure_2d=False, dtype='numeric')
+        y = np.asarray(y)
```
I am not sure this is the right things to do. Now passing a sparse matrix will give something funky and will error, while the `check_array` was used to filter this part (in addition of checking for non-finite values).

I think the safe thing to do here is to allow for nd array using `allow_nd=True` in `check_array`.

glemaitre commented on sklearn/compose/tests/test_target.py:
```diff
@@ -198,6 +198,25 @@ def test_transform_target_regressor_2d_transformer_multioutput():
 
 
+def test_transform_target_regressor_3d_target():
+    # Check with a 3D target with a transformer that reshapes the target
```
You can add:

```python
# Non-regression test for:
# https://github.com/scikit-learn/scikit-learn/issues/18866
```

glemaitre reviewed (commented):
I think that we should still keep the `check_array` as a safe guard for the moment. We can style accomodate your use case by allowing `nd` array in `check_array`. We can later see if we should do less validation and delegate the `y` validation to the transformer itself.

Please add an entry to the change log at `doc/whats_new/v*.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

cmarmo commented:
Hi @panangam would you be able to address the comments from the reviewer? Thanks! Feel free to ask if you need some clarifications.

panangam committed:
add a comment to reference issue

panangam committed:
bring check_array back but set allow_nd=True

panangam committed:
add an entry to changelog

panangam committed:
Merge branch 'main' into transformedtargetregressor-accept-nd

panangam commented:
@glemaitre @cmarmo I know it's been forever but I've finally addressed the comment, bringing `check_array` back and did the book keeping stuff (commenting, changelogs).

cmarmo commented:
Hi @panangam thanks for your updates. 0.24 has been released in December 2020, do you mind moving your changelog in `doc/whats_new/v1.0.rst`? Thanks!

panangam commented:
> Hi @panangam thanks for your updates. 0.24 has been released in December 2020, do you mind moving your changelog in `doc/whats_new/v1.0.rst`? Thanks!

Oh whoa, we're getting to 1.0? Will do!

panangam committed:
move changelog entry for compose.TransformedTargetRegressor from v0.24 to v1.0

glemaitre committed:
Merge branch 'main' into transformedtargetregressor-accept-nd

glemaitre reviewed (approved):
LGTM

thomasjpfan reviewed (approved):
Thank you for the PR @panangam !

LGTM
